src/main/java/io/vavr/collection/Iterator.java

```diff
@@ -1722,10 +1722,16 @@ public U next() {
         if (!hasNext()) {
             return Tuple.of(empty(), empty());
         } else {
-            final Stream<T> that = Stream.ofAll(this);
-            final Iterator<T> first = that.iterator().filter(predicate);
-            final Iterator<T> second = that.iterator().filter(predicate.negate());
-            return Tuple.of(first, second);
+            Stream<T> first = Stream.empty();
+            Stream<T> second = Stream.empty();
+            for (T t : this) {
+                if (predicate.test(t)) {
+                    first = first.append(t);
```

Comment:
As far as I remember, Stream.append is an O(n) operation, so the whole op will be O(n^2). 
I believe you should use two java's ArrayList objects to partition.

Reply:
Thanks for your pointer @ruslansennov . This should be fixed now.